buildchain/buildchain/targets/serialize.py

```diff
@@ -95,3 +121,64 @@ def _render(self) -> Callable[[Any, Path], None]:
     def _run(self) -> None:
         """Render the file."""
         self._render(self._data, self._dest)
+
+
+# YAML {{{
+
+
+class YAMLDocument():
+    """A YAML document, with an optional preamble (like a shebang)."""
+    class Literal(str):
+        """A large block of text, to be rendered as a block scalar."""
+
+    class ByteString(bytes):
+        """A binary string, to be rendered as a base64-encoded literal."""
+
+    @classmethod
+    def text(cls, value: str) -> 'YAMLDocument.Literal':
+        """Cast the value to a Literal."""
+        return cls.Literal(value)
+
+    @classmethod
+    def bytestring(cls, value: bytes) -> 'YAMLDocument.ByteString':
+        """Cast the value to a ByteString."""
+        return cls.ByteString(value)
+
+
+SaltState = collections.namedtuple(
+    'SaltState', ['content', 'shebang', 'imports']
+)
+
+
+def _literal_representer(dumper: yaml.BaseDumper, data: Any) -> Any:
+    scalar = yaml.representer.SafeRepresenter.represent_str( # type: ignore
+        dumper, data
+    )
+    scalar.style = '|'
+    return scalar
+
+
+def _bytestring_representer(dumper: yaml.BaseDumper, data: Any) -> Any:
+    return _literal_representer(
+        dumper, base64.encodebytes(data).decode('utf-8')
+    )
+
+
+def _yaml_dump(data: Sequence[Any], fp: IO[Any]) -> None:
+    dumper = yaml.SafeDumper(fp, sort_keys=False) # type: ignore
+    dumper.add_representer( # type: ignore
+        YAMLDocument.Literal, _literal_representer
+    )
+    dumper.add_representer( # type: ignore
+        YAMLDocument.ByteString, _bytestring_representer
+    )
+    try:
+        dumper.open() # type: ignore
+        for document in data:
+            dumper.represent(document) # type: ignore
```

Comment:
I assume this then adds the `---`?

Reply:
Yes, I've tested locally.

And the code is highly inspired from [official PyYAML code](https://github.com/yaml/pyyaml/blob/master/lib/yaml/__init__.py#L278-L284).